crates/breez-sdk/core/src/sync/spark.rs

```diff
@@ -0,0 +1,221 @@
+            // we can fetch all in one go using same batch size
+            let parent_transactions = self
+                .spark_wallet
+                .list_token_transactions(ListTokenTransactionsRequest {
+                    paging: Some(PagingFilter::new(None, Some(PAYMENT_SYNC_BATCH_SIZE), None)),
+                    owner_public_keys: Some(Vec::new()),
+                    token_transaction_hashes: token_transactions_prevout_hashes,
+                    ..Default::default()
+                })
+                .await?;
+
+            info!(
+                "Syncing token payments to storage, offset = {next_offset}, transactions = {}",
+                token_transactions.len()
+            );
+            // Process transfers in this page
+            for transaction in &token_transactions {
+                let tx_inputs_are_ours = match &transaction.inputs {
+                    spark_wallet::TokenInputs::Transfer(token_transfer_input) => {
+                        let first_input = token_transfer_input.outputs_to_spend.first().ok_or(
+                            SdkError::Generic("No input in token transfer input".to_string()),
+                        )?;
+                        let parent_transaction = parent_transactions
+                            .iter()
+                            .find(|tx| tx.hash == first_input.prev_token_tx_hash)
+                            .ok_or(SdkError::Generic(
+                                "Parent transaction not found".to_string(),
+                            ))?;
+                        let output = parent_transaction
+                            .outputs
+                            .get(first_input.prev_token_tx_vout as usize)
+                            .ok_or(SdkError::Generic("Output not found".to_string()))?;
+                        output.owner_public_key == our_public_key
+                    }
+                    spark_wallet::TokenInputs::Mint(_) | spark_wallet::TokenInputs::Create(_) => {
+                        false
+                    }
+                };
+
+                // Create payment records
+                let payments = token_transaction_to_payments(
+                    &self.spark_wallet,
+                    transaction,
+                    tx_inputs_are_ours,
+                )
+                .await?;
+
+                for payment in payments {
+                    // Stop syncing if we encounter a finalized payment that we have already processed
+                    if let Ok(Payment {
+                        status: PaymentStatus::Completed | PaymentStatus::Failed,
+                        ..
+                    }) = self.storage.get_payment_by_id(payment.id.clone()).await
+                    {
+                        info!(
+                            "Encountered already finalized payment {}, stopping sync",
+                            payment.id
+                        );
+                        break 'page_loop;
```

Comment:
I think there might be a problem here if we started sync, got 1-2 pages and then restarted.
If before we restart some payments were done then the sync will fetch some payments that exists in the db and according to this logic may exit early and not sync all the way back.
One way I think of solving this is to have the newer payment id that we should sync to (and then stop).
One way to solve this would be:
* maintain a payment id (in a separate cached item) that is the newest id we should sync up to (stop when we getting into this payment).
* Whenever we reach that payment we update it to be the newest final payment id in the db.
* On first sync is is None so we should sync until we have no more pages and then updated it to the newest final one.